[io] add test bed for streaming large objects

jblomer opened the pull request.

ferdymercury committed:
[io] Properly abort when buffer size overflows max integer

Fixes https://github.com/root-project/root/issues/14770

[io] add more checks in TBuffer functions

as suggested by jblomer

pcanal committed:
io: Add data structure for 64 bit byte count.

In order to support 64 bits byte counts, which do not fit in the space reserved
for them in the stream (4 bytes minus 2 control bits) and to work around the
fact that the variables that holds the position and the byte count information in
Streamer functions are only 32 bits (see arguments to TBufferFile::WriteVersion
and TBufferFile::ReadVersion), we need to pass them indirectly when they needs 64 bits

Since the streaming is inherently serial, we can leverage the
sequence of calls and cache the 64 bits values in a queue.

The bytecount that can not be stored in place inside the io stream will be held
in a collection (fByteCounts) that need to be stored externally to the
buffer.

pcanal committed:
io: Support 64 bits position/byte-count in writing

pcanal committed:
io: Support 64 bits position/byte-count in reading

pcanal committed:
io/nfc: add comment

jblomer committed:
[io] add unit test infrastructure for streaming large objects

Uses RStreamerField as a test bed for (de-)serializing large objects
with TBufferFile.

silverweed commented on io/io/src/TBufferFile.cxx:
```diff
@@ -323,8 +323,13 @@ void TBufferFile::SetByteCount(ULong64_t cntpos, Bool_t packInVersion)
+
    char  *buf = (char *)(fBuffer + cntpos);
+   if ((fBufCur - fBuffer - cntpos - sizeof(UInt_t)) >= kMaxMapCount) {
+      tobuf(buf, 0 | kByteCountMask);
```
The reason to write `0 | kByteCountMask` is clear for us (today) but it might be confusing to anyone not familiar with the idea behind the plan; maybe a comment would help here.

pcanal replied:
On the other hand, this code is already almost obsolete, i.e. replaced by https://github.com/root-project/root/pull/20574 (which is still missing a few wrinkles: decision on Reset behavior, getter and setter)

pcanal committed:
io: Add Set/GetByteCounts to TBufferFile

Note to store and restore the larger than 1GB byte count use:

  // Copy the content of the const reference.
  auto bytecounts{b.GetByteCounts()};
  ...
  b.SetByteCounts(std::move(bytecounts));

pcanal committed:
io: Add reset of new bytecount infrastructure

pcanal committed:
io: Add new bytecount test

pcanal committed:
io: Suppress error message about large byte count

pcanal committed:
[NFC] io white space / byte count related

pcanal committed:
NFC io: fix doc typo

pcanal committed:
io: doc clarifications - bytecount related

pcanal committed:
io: Clarify byte count related limits

pcanal committed:
io: ReadVersion should always take both position and byte count

pcanal committed:
io: Correct handling of byte count pos in TBufferFile::ReadObjectAny

pcanal committed:
io: compile testLargeByteCounts

pcanal committed:
io: Fix up clarify byte count related limits

pcanal committed:
io: In TBufferFile::ReadClass fix recording of byte count pos.

This fixes TBufferFile::ReadObjectAny handling of long range byte counts

pcanal committed:
io: Add missing CheckByteCount in TStreamerElements.

Those were essentially replaced by a 'set the cursor to the end'
call in commit 7d25b75b8877a03ee579e97d91ee8f719721d6d5.
It is dubious whether these changes are needed or were an heurestic
to ignore (temporarily !?) race conditions.

pcanal committed:
io/roofit: Correct CheckByteCount placement.

Due to the need for using a stack of bytecount position, we can no longer support redundant
calls to CheckByteCount.

pcanal committed:
NFC io: tighten local variable definition

pcanal committed:
io: CheckByteCount need class pointer/name as argument

pcanal committed:
io: Add debugging utility for mismatch Set/CheckByteCount

pcanal committed:
io/tree: Remove redundant CheckByteCount

pcanal committed:
io: byte count stack handling in skip object

pcanal committed:
io: Add TBuffer::ByteCountWriter RAII.

This simplify the implementation of calling `ReserveByteCount`
and then `SetByteCount` and could also simplify removing the
use of the ByteCountStack in some/most cases.

Note: `ReserveByteCount` is protected and the existing code
that does the same semantic action was only recording the position
in a 32 bits integer.  That code needs to be updated by either
making `ReserverByteCount` public (making removal of external use
of the ByteCountStack more difficult) or by introducing a RAII
that hides the size of the integer.

pcanal committed:
io: Upgrade TArray::WriteArray to support long range byte count

pcanal committed:
roottest: Fix cmake macro used outside of roottest dir.

In order to allow using the roottest CMake macro outside of the roottest sub-directory,
we need to make the variable used by those macros 'cache variable' so that they can use
elsewhere.

This is a hack and works only after the second CMake configuration/invocation.

pcanal committed:
cmake: Include VStudio in the `Ninja` `RESOURCE_LOCK`.

When using the Microsoft Visual Studio generator we have the exact same problem,
that it might trigger a rebuild of ROOT so we need to apply the same solution.

See 06e00a270e6e4f1157bfc344ad0dcdb6c45bb01a.

PS. We might want to eventually rename the resource lock.

pcanal committed:
io-test: disable large byte count test on 32bits platforms

pcanal committed:
cmake: Use single variable to select Ninja Build resource lock

pcanal committed:
cmake: Rename Ninja Build resource lock to CMake Build.

It is need by at least Ninja and the Visual Studio generator

pcanal committed:
cmake: Add missing config to roottest (MSVC) (re)build

pcanal committed:
io: Extend code documentation related to ByteCount

pcanal committed:
NFC: white space

jblomer commented:
Replaced by #22430
